File: iOS/Mobile/DirectoryViewController.m
//
//  DirectoryViewController.m
//  Mobile
//
//  Created by Jason Hocker on 10/4/12.
//  Copyright (c) 2012-2014 Ellucian. All rights reserved.
//

#import <AddressBook/AddressBook.h>
#import "DirectoryViewController.h"
#import "DirectoryEntry.h"
#import "CurrentUser.h"
#import "Base64.h"
#import "AppDelegate.h"
#import "DirectoryEntryViewController.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"
#import "AppearanceChanger.h"
#import "Ellucian_GO-Swift.h"

@interface DirectoryViewController ()

@property (nonatomic, strong) NSURLConnection *connection;
@property (nonatomic, strong) NSMutableData *data;
@property (nonatomic, assign) ABPersonSortOrdering sort;
@property (nonatomic, assign) SEL sortSelector;
@property (nonatomic, assign) UILocalizedIndexedCollation *collation;
@property (readwrite, copy, nonatomic) NSArray *tableData;
@property (nonatomic, assign) BOOL firstNameFirst;
@property (nonatomic, assign) DirectoryViewType selectedScope;
@property (nonatomic, strong) NSArray *scopesType;
@end

@implementation DirectoryViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    ABAddressBookRef addressBook;
    addressBook = ABAddressBookCreateWithOptions(NULL, NULL);
    __block BOOL accessGranted = NO;
    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
    ABAddressBookRequestAccessWithCompletion(addressBook, ^(bool granted, CFErrorRef error) {
        accessGranted = granted;
        dispatch_semaphore_signal(sema);
    });
    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
    
    if(accessGranted) {
        self.sort = ABPersonGetSortOrdering();
        self.sortSelector = self.sort == kABPersonCompositeNameFormatFirstNameFirst ? @selector(firstNameSort) : @selector(lastNameSort);
    
        self.firstNameFirst = ABPersonGetCompositeNameFormatForRecord(NULL) == kABPersonCompositeNameFormatFirstNameFirst;
        CFRelease(addressBook);
    } else {
        self.sort = kABPersonCompositeNameFormatFirstNameFirst;
        self.sortSelector = @selector(lastNameSort);
        self.firstNameFirst = YES;
    }


    self.title = self.module.name;
    self.tableData = [self partitionObjects:[[NSArray alloc] init] collationStringSelector:self.sortSelector];
    
    if(self.initialScope) {
        self.selectedScope = self.initialScope;
    }
    
    NSMutableArray *mutableScopesTitle = [[NSMutableArray alloc] init];
    NSMutableArray *mutableScopesTypes = [[NSMutableArray alloc] init];
    
    //Directory module
    if(self.module && [self.module.type isEqualToString:@"directory"]) {
        if([[self.module propertyForKey:@"student"] isEqualToString:@"true"] && !self.hideStudents) {
            if(!self.selectedScope) self.selectedScope = DirectoryViewTypeStudent;
            [mutableScopesTitle addObject:NSLocalizedString(@"Students", @"student search scope in directory")];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeStudent]];
        }
        if([[self.module propertyForKey:@"faculty"] isEqualToString:@"true"] && !self.hideFaculty) {
            if(!self.selectedScope) self.selectedScope = DirectoryViewTypeFaculty;
            [mutableScopesTitle addObject:NSLocalizedString(@"Faculty/Staff", @"facilty/staff search scope in directory") ];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeFaculty]];
        }
        if([mutableScopesTitle count] > 1) {
            [mutableScopesTitle addObject:NSLocalizedString(@"All", @"all search scope in directory") ];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeAll]];
        }
    } else { //not a directory module, but reusing search
        NSUserDefaults *defaults = [AppGroupUtilities userDefaults];
        if([defaults objectForKey:@"urls-directory-studentSearch"] && !self.hideStudents) {
            if(!self.selectedScope) self.selectedScope = DirectoryViewTypeStudent;
            [mutableScopesTitle addObject:NSLocalizedString(@"Students", @"student search scope in directory") ];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeStudent]];
        }
        if([defaults objectForKey:@"urls-directory-facultySearch"] && !self.hideFaculty ) {
            if(!self.selectedScope) self.selectedScope = DirectoryViewTypeFaculty;
            [mutableScopesTitle addObject:NSLocalizedString(@"Faculty/Staff", @"facilty/staff search scope in directory") ];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeFaculty]];
        }
        if([defaults objectForKey:@"urls-directory-allSearch"] && [mutableScopesTitle count] > 1) {
            if(!self.selectedScope) self.selectedScope = DirectoryViewTypeAll;
            [mutableScopesTitle addObject:NSLocalizedString(@"All", @"all search scope in directory") ];
            [mutableScopesTypes addObject:[NSNumber numberWithInt:DirectoryViewTypeAll]];
        }
    }
    
    self.scopesType = [mutableScopesTypes copy];
    if([mutableScopesTitle count] > 1) {
        self.searchBar.scopeButtonTitles = [mutableScopesTitle copy];

        self.searchBar.tintColor = [UIColor primaryColor];

        if(self.initialScope) {
            self.searchBar.selectedScopeButtonIndex = [self.scopesType indexOfObject:[NSNumber numberWithInt:self.initialScope]];
        } else {
            self.searchBar.selectedScopeButtonIndex = 0;
        }
        self.searchBar.showsScopeBar = YES;
        [self.searchBar sizeToFit];
    } else {
        self.searchBar.scopeButtonTitles = nil;
        self.searchBar.showsScopeBar = NO;
        [self.searchBar sizeToFit];
    }
  
    if(self.initialQueryString) {
        self.searchBar.text = self.initialQueryString;
        [self search:self.initialQueryString];
    }
}


- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [self sendView:@"Directory page" forModuleNamed:self.module.name];
    
    if(!self.initialQueryString) {
        [self.searchBar becomeFirstResponder];
    }
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [self.searchBar resignFirstResponder];
    [self clear];
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    [self search:searchBar.text];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    [searchBar resignFirstResponder];
}

- (void)connection:(NSURLConnection *)theConnection didReceiveData:(NSData *)incrementalData {
    if (self.data==nil) {
        self.data = [[NSMutableData alloc] initWithCapacity:2048];
    }
    [self.data appendData:incrementalData];
    
    
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

- (void)connectionDidFinishLoading:(NSURLConnection*)theConnection {
    [self parseResponse:self.data];
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    self.data = nil;
    self.connection = nil;

}

- (void) parseResponse:(NSData *)responseData
{
    NSMutableArray *tempContacts = [[NSMutableArray alloc] init];
    NSError *error;
    if(responseData)
    {
        
        NSDictionary* json = [NSJSONSerialization
                              JSONObjectWithData:responseData
                              options:kNilOptions
                              error:&error];
        
        //create/update objects
        for(NSDictionary *entry in [json objectForKey:@"entries"]) {
            DirectoryEntry *directoryEntry = [[DirectoryEntry alloc] init];

            if([entry objectForKey:@"personId"] != [NSNull null]) {
                directoryEntry.personId = [entry objectForKey:@"personId"];
            }
            if([entry objectForKey:@"username"] != [NSNull null]) {
                directoryEntry.username = [entry objectForKey:@"username"];
            }
            if([entry objectForKey:@"displayName"] != [NSNull null]) {
                directoryEntry.displayName = [entry objectForKey:@"displayName"];
            }
            if([entry objectForKey:@"firstName"] != [NSNull null]) {
                directoryEntry.firstName = [entry objectForKey:@"firstName"];
            }
            if([entry objectForKey:@"middleName"] != [NSNull null]) {
                directoryEntry.middleName = [entry objectForKey:@"middleName"];
            }
            if([entry objectForKey:@"lastName"] != [NSNull null]) {
                directoryEntry.lastName = [entry objectForKey:@"lastName"];
            }
            if([entry objectForKey:@"title"] != [NSNull null]) {
                directoryEntry.title = [entry objectForKey:@"title"];
            }
            if([entry objectForKey:@"office"] != [NSNull null]) {
                directoryEntry.office = [entry objectForKey:@"office"];
            }
            if([entry objectForKey:@"department"] != [NSNull null]) {
                directoryEntry.department = [entry objectForKey:@"department"];
            }
            if([entry objectForKey:@"phone"] != [NSNull null]) {
                directoryEntry.phone = [entry objectForKey:@"phone"];
            }
            if([entry objectForKey:@"mobile"] != [NSNull null]) {
                directoryEntry.mobile = [entry objectForKey:@"mobile"];
            }
            if([entry objectForKey:@"email"] != [NSNull null]) {
                directoryEntry.email = [entry objectForKey:@"email"];
            }
            if([entry objectForKey:@"street"] != [NSNull null]) {
                directoryEntry.street = [[entry objectForKey:@"street"] stringByReplacingOccurrencesOfString:@"\\n" withString:@"\n"];
            }
            if([entry objectForKey:@"room"] != [NSNull null]) {
                directoryEntry.room = [entry objectForKey:@"room"];
            }
            if([entry objectForKey:@"postOfficeBox"] != [NSNull null]) {
                directoryEntry.postOfficeBox = [entry objectForKey:@"postOfficeBox"];
            }
            if([entry objectForKey:@"city"] != [NSNull null]) {
                directoryEntry.city = [entry objectForKey:@"city"];
            }
            if([entry objectForKey:@"state"] != [NSNull null]) {
                directoryEntry.state = [entry objectForKey:@"state"];
            }
            if([entry objectForKey:@"postalCode"] != [NSNull null]) {
                directoryEntry.postalCode = [entry objectForKey:@"postalCode"];
            }
            if([entry objectForKey:@"country"] != [NSNull null]) {
                directoryEntry.country = [entry objectForKey:@"country"];
            }
            if([entry objectForKey:@"prefix"] != [NSNull null]) {
                directoryEntry.prefix = [entry objectForKey:@"prefix"];
            }
            if([entry objectForKey:@"suffix"] != [NSNull null]) {
                directoryEntry.suffix = [entry objectForKey:@"suffix"];
            }
            [tempContacts addObject:directoryEntry];
        }
    }
    
    if(self.sort == kABPersonSortByFirstName) {
        [tempContacts sortUsingDescriptors:[NSArray arrayWithObject: [NSSortDescriptor sortDescriptorWithKey:@"firstName" ascending:YES]]];
    } else {
        [tempContacts sortUsingDescriptors:[NSArray arrayWithObject: [NSSortDescriptor sortDescriptorWithKey:@"lastName" ascending:YES]]];
    }

    self.tableData = [self partitionObjects:[tempContacts copy] collationStringSelector:self.sortSelector];
    [self.tableView reloadData];
}



- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    DirectoryEntry *entry = nil;
    entry = [[self.tableData objectAtIndex:indexPath.section] objectAtIndex:indexPath.row];

    static NSString *CellIdentifier = @"Directory Name Cell";

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryNone;
    }

    UILabel *textLabel = (UILabel *)[cell viewWithTag:1];
    textLabel.text = [entry nameOrderedByFirstName:self.firstNameFirst];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    DirectoryEntry *entry = [[self.tableData objectAtIndex:indexPath.section] objectAtIndex:indexPath.row];
    [self performSegueWithIdentifier:@"Show Directory Profile" sender:entry];
}

- (void)searchBar:(UISearchBar *)searchBar selectedScopeButtonIndexDidChange:(NSInteger)selectedScope
{
    [self sendEventWithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionButton_Press withLabel:@"Select directory type" withValue:nil forModuleNamed:self.module.name];
    self.selectedScope = [[self.scopesType objectAtIndex:self.searchBar.selectedScopeButtonIndex] intValue];
    if(searchBar.text.length > 0) {
        [self search:searchBar.text];
    }
}

-(void)search:(NSString *)searchString
{
    self.initialQueryString = searchString;
    if([searchString length] > 0) {
        NSString *url = nil;
        if(self.module && [self.module.type isEqualToString:@"directory"]) {
            if(self.selectedScope == DirectoryViewTypeStudent) {
                url = [self.module propertyForKey:@"studentSearch"];
            } else if(self.selectedScope == DirectoryViewTypeFaculty) {
                url = [self.module propertyForKey:@"facultySearch"];
            } else {
                 url = [self.module propertyForKey:@"allSearch"];
            }
        } else {
            NSUserDefaults *defaults = [AppGroupUtilities userDefaults];
            if(self.selectedScope == DirectoryViewTypeStudent) {
                url = [defaults objectForKey:@"urls-directory-studentSearch"];
            } else if(self.selectedScope == DirectoryViewTypeFaculty) {
               url = [defaults objectForKey:@"urls-directory-facultySearch"];
            } else {
               url = [defaults objectForKey:@"urls-directory-allSearch"];
            }
        }
        url = [NSString stringWithFormat:@"%@?searchString=%@", url, [searchString stringByAddingPercentEscapesUsingEncoding: NSUTF8StringEncoding]];

        CurrentUser *user = [CurrentUser sharedInstance];
        NSMutableString *loginString = (NSMutableString*)[@"" stringByAppendingFormat:@"%@:%@", user.userauth, user.getPassword];

        NSString *encodedLoginData = [Base64 encode:[loginString dataUsingEncoding:NSUTF8StringEncoding]];

        NSString *authHeader = [@"Basic " stringByAppendingFormat:@"%@", encodedLoginData];
        
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url] cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:30.0];
        [request setValue:authHeader forHTTPHeaderField:@"Authorization"];

        NSLog(@"%@",url);
        if(self.connection!=nil){ //cancel if in process
            [self.connection cancel];
        }
        self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
    } else {
        [self clear];
    }
}

-(NSArray *)partitionObjects:(NSArray *)array collationStringSelector:(SEL)selector

{
    UILocalizedIndexedCollation *collation = [UILocalizedIndexedCollation currentCollation];
    
    NSInteger sectionCount = [[collation sectionTitles] count]; //section count is take from sectionTitles and not sectionIndexTitles
    NSMutableArray *unsortedSections = [NSMutableArray arrayWithCapacity:sectionCount];
    
    //create an array to hold the data for each section
    for(int i = 0; i < sectionCount; i++)
    {
        [unsortedSections addObject:[NSMutableArray array]];
    }
    
    //put each object into a section
    for (id object in array)
    {
        NSInteger index = [collation sectionForObject:object collationStringSelector:selector];
        [[unsortedSections objectAtIndex:index] addObject:object];
    }
    
    NSMutableArray *sections = [NSMutableArray arrayWithCapacity:sectionCount];
    
    //sort each section
    for (NSMutableArray *section in unsortedSections)
    {
        NSArray *sorted = [collation sortedArrayFromArray:section collationStringSelector:selector];
        [sections addObject:sorted];
    }
    
    return sections;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [[self.tableData objectAtIndex:section] count];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [[[UILocalizedIndexedCollation currentCollation] sectionTitles] count];
}

- (NSString *)tableView:(UITableView *)tableView stringForTitleForHeaderInSection:(NSInteger)section
{
    BOOL showSection = [[self.tableData objectAtIndex:section] count] != 0;
    //only show the section title if there are rows in the section
    return (showSection) ? [[[UILocalizedIndexedCollation currentCollation] sectionTitles] objectAtIndex:section] : nil;
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index
{
    return [[UILocalizedIndexedCollation currentCollation] sectionForSectionIndexTitleAtIndex:index];
}

-(void) clear
{
    self.tableData = [self partitionObjects:[[NSArray alloc] init] collationStringSelector:self.sortSelector];
    [self.tableView reloadData];

}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    
    if ([[segue identifier] isEqualToString:@"Show Directory Profile"])
    {
        DirectoryEntryViewController *vc = (DirectoryEntryViewController *)[segue destinationViewController];
        vc.entry = sender;
        vc.module = self.module;
    }
}

@end

File: iOS/Mobile/DirectoryViewController.swift
    //
    //  DirectoryViewController.swift
    //  Mobile
    //
    //  Created by Jason Hocker on 12/2/15.
    //  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
    //
    
    import Foundation
    
    class DirectoryViewController : UITableViewController, UISearchBarDelegate, DirectoryFilterDelegate {
        
        let searchDelayInterval : NSTimeInterval = 0.5
        
        @IBOutlet var searchBar: UISearchBar!
        var module : Module?
        var hideStudents : Bool?
        var hideFaculty : Bool?
        
        var forcedFilteredGroup : String?
        var searchDelayer : NSTimer?
        var legacySearch = false
        var connection : NSURLConnection?
        var hiddenGroups = [String]()
        
        @IBOutlet var filterButton: UIBarButtonItem!
        var tableData = [[DirectoryEntry]]()
        var entries = [DirectoryEntry]()
        var data : NSMutableData?
        var groups = [DirectoryDefinitionProtocol]()
        var inSearch = false
        var sortByLastName = true
        
        override func viewDidLoad() {
            super.viewDidLoad()
            
            if entries.count > 0 {
                //coming from another source, such as course roster
                inSearch = true
                self.tableData = self.partitionObjects(self.entries, collationStringSelector: Selector("lastNameSort"))
            }
            
            buildGroups()
            
            tableView.rowHeight = UITableViewAutomaticDimension
            tableView.estimatedRowHeight = 53
            
            self.title = self.module?.name
            
            if UIScreen.mainScreen().traitCollection.userInterfaceIdiom == .Pad {
                self.splitViewController?.preferredDisplayMode = .AllVisible;
            } else {
                self.splitViewController?.preferredDisplayMode = .Automatic;
            }
            
            if groups.count <= 1 {
                filterButton.enabled = false;
                self.navigationController?.navigationBar.topItem!.rightBarButtonItem = nil;
            }
            
            if let searchBar = searchBar {
                searchBar.becomeFirstResponder()
            }
            
            
            NSNotificationCenter.defaultCenter().addObserver(self, selector: "signInHappened", name: kLoginExecutorSuccess, object: nil)
        }
        
        override func viewDidAppear(animated: Bool) {
            super.viewDidAppear(animated)
            sendView("Directory page", forModuleNamed: self.module?.name)
            
        }
        
        func doDelayedSearch(timer: NSTimer) {
            searchDelayer = nil
            doSearch()
        }
        
        func doSearch() {
            if !self.legacySearch {
                doModernSearch()
            } else {
                doLegacySearch()
            }
        }
        
        func doLegacySearch() {
            if let searchString = self.searchBar.text {
                if searchString.characters.count > 0 {
                    var url: String? = nil
                    if let module = module where self.module?.type == "directory" {
                        
                        let groupsInUse = self.groups.filter() {
                            !self.hiddenGroups.contains($0.internalName!)
                        }
                        let filteredGroups = groupsInUse.map() {
                            $0.internalName!
                        }
                        
                        if filteredGroups.count == 0 {
                            clear()
                            return
                        } else if filteredGroups.count == 2 {
                            url = module.propertyForKey("allSearch")
                        } else if filteredGroups.count == 1 && filteredGroups.contains("student") {
                            url = module.propertyForKey("studentSearch")
                        } else if filteredGroups.count == 1 && filteredGroups.contains("faculty") {
                            url = module.propertyForKey("facultySearch")
                        }
                    }
                    
                    let encodedSearchString = searchString.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
                    url = "\(url!)?searchString=\(encodedSearchString!)"
                    
                    let request = NSMutableURLRequest(URL: NSURL(string: url!)!, cachePolicy: .UseProtocolCachePolicy, timeoutInterval: 30.0)
                    let authenticationMode = AppGroupUtilities.userDefaults()?.objectForKey("login-authenticationType") as? String
                    if authenticationMode == nil || authenticationMode == "native" {
                        request.addAuthenticationHeader()
                    }
                    
                    if let connection = self.connection {
                        connection.cancel()
                    }
                    self.connection = NSURLConnection(request: request, delegate: self)
                    UIApplication.sharedApplication().networkActivityIndicatorVisible = true
                    
                }
            } else {
                clear()
            }
        }
        
        func doModernSearch() {
            if let searchString = self.searchBar.text {
                if searchString.characters.count > 0 {
                    var url = module?.propertyForKey("baseSearch")
                    
                    var groupsInUse = self.groups
                    if self.hiddenGroups.count > 0 {
                        groupsInUse = self.groups.filter() {
                            !self.hiddenGroups.contains($0.internalName!)
                        }
                    }
                    
                    let filteredGroups = groupsInUse.map() {
                        return $0.internalName as String!
                        } as [String]
                    
                    if filteredGroups.count > 0 {
                        
                        let encodedSearchString = searchString.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
                        let encodedDirectories = filteredGroups.joinWithSeparator(",").stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
                        url = "\(url!)?searchString=\(encodedSearchString!)&directories=\(encodedDirectories!)"
                        let request = NSMutableURLRequest(URL: NSURL(string: url!)!, cachePolicy: .UseProtocolCachePolicy, timeoutInterval: 30.0)
                        if CurrentUser.sharedInstance().isLoggedIn {
                            request.addAuthenticationHeader()
                        }
                        if let connection = self.connection {
                            connection.cancel()
                        }
                        self.connection = NSURLConnection(request: request, delegate: self)
                        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
                    } else {
                        clear()
                    }
                }
            } else {
                clear()
            }
        }
        
        func clear() {
            if !self.splitViewController!.collapsed {
                self.performSegueWithIdentifier("Empty Entry", sender: nil)
            }
            self.tableData = self.partitionObjects([DirectoryEntry](), collationStringSelector: Selector("lastNameSort"))
            self.tableView.reloadData()
            inSearch = false
        }
        
        // MARK: UISearchBarDelegate
        
        func searchBar(searchBar: UISearchBar, textDidChange searchText: String) {
            if let searchDelayer = searchDelayer {
                searchDelayer.invalidate()
            }
            if searchText.characters.count > 0 {
                searchDelayer = NSTimer.scheduledTimerWithTimeInterval(searchDelayInterval, target: self, selector: Selector("doDelayedSearch:"), userInfo: searchText, repeats: false)
                inSearch = true
            } else {
                clear()
            }
        }
        
        func searchBarSearchButtonClicked(searchBar: UISearchBar) {
            searchBar.resignFirstResponder()
            searchDelayer = nil
            doSearch()
        }
        
        //MARK: segue
        override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
            if segue.identifier == "Show Filter" {
                self.sendEventWithCategory(kAnalyticsCategoryUI_Action, withAction: kAnalyticsActionButton_Press, withLabel: "Select directory type", withValue: nil, forModuleNamed: self.module?.name)
                let navigationController = segue.destinationViewController as! UINavigationController
                let detailController = navigationController.viewControllers[0] as! DirectoryFilterViewController
                detailController.module = self.module
                detailController.hiddenGroups = self.hiddenGroups
                detailController.groups = self.groups
                detailController.delegate = self
            } else if segue.identifier == "Show Directory Profile" {
                let detailController : DirectoryEntryViewController
                if segue.destinationViewController is DirectoryEntryViewController {
                    detailController = segue.destinationViewController as! DirectoryEntryViewController
                } else {
                    let navigationController = segue.destinationViewController as! UINavigationController
                    detailController = navigationController.viewControllers[0] as! DirectoryEntryViewController
                }
                detailController.entry = sender as? DirectoryEntry;
                detailController.module = self.module;
            }
        }
        
        //MARK: other
        func partitionObjects(array: [DirectoryEntry], collationStringSelector selector: Selector) -> [[DirectoryEntry]] {
            let collation: UILocalizedIndexedCollation = UILocalizedIndexedCollation.currentCollation()
            let sectionCount = collation.sectionTitles.count
            //section count is take from sectionTitles and not sectionIndexTitles
            var unsortedSections = [[DirectoryEntry]](count: sectionCount, repeatedValue:[DirectoryEntry]())
            
            //put each object into a section
            for object in array {
                let index = collation.sectionForObject(object, collationStringSelector: selector)
                unsortedSections[index].append(object)
            }
            var sections = [[DirectoryEntry]](count: sectionCount, repeatedValue:[DirectoryEntry]())
            //sort each section
            for i in 0 ..< unsortedSections.count {
                let sorted = collation.sortedArrayFromArray(unsortedSections[i], collationStringSelector: selector) as! [DirectoryEntry]
                sections[i] = sorted
            }
            
            return sections;
        }
        
        func parseResponse(responseData: NSData) {
            entries = DirectoryEntry.parseResponse(responseData)
            buildTableData(entries)
            
        }
        
        func connection(theConnection: NSURLConnection, didReceiveData incrementalData: NSData) {
            if self.data == nil {
                self.data = NSMutableData(capacity: 2048)
            }
            self.data?.appendData(incrementalData)
            
        }
        
        func connectionDidFinishLoading(theConnection: NSURLConnection) {
            if let data = self.data {
                self.parseResponse(data)
            }
            UIApplication.sharedApplication().networkActivityIndicatorVisible = false
            self.data = nil
            self.connection = nil
        }
        
        func buildTableData(entries: [DirectoryEntry]) {
            
            var error: Unmanaged<CFError>?
            let addressBook: ABAddressBook? = ABAddressBookCreateWithOptions(nil, &error)?.takeRetainedValue()
            if let addressBook = addressBook {
                // request permission to use it
                ABAddressBookRequestAccessWithCompletion(addressBook) {
                    granted, error in
                    
                    if !granted {
                        self.sortByLastName = true
                        self.tableData = self.partitionObjects(self.entries, collationStringSelector: Selector("lastNameSort"))
                        dispatch_async(dispatch_get_main_queue(),{
                            self.tableView.reloadData()
                            
                        })
                    } else {
                        
                        let sort = Int(ABPersonGetSortOrdering())
                        if sort == Int(kABPersonSortByFirstName) {
                            self.sortByLastName = false
                            self.tableData = self.partitionObjects(self.entries, collationStringSelector: Selector("firstNameSort"))
                        }
                        else {
                            self.sortByLastName = true
                            self.tableData = self.partitionObjects(self.entries, collationStringSelector: Selector("lastNameSort"))
                        }
                        dispatch_async(dispatch_get_main_queue(),{
                            self.tableView.reloadData()
                        })
                    }
                }
            } else {
                self.sortByLastName = true
                self.tableData = self.partitionObjects(self.entries, collationStringSelector: Selector("lastNameSort"))
                self.tableView.reloadData()
            }
        }
        
        override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
            if section == 0 {
                var count = 0
                if inSearch && entries.count == 0 || searchBar?.text?.characters.count == 0 {
                    count += 1
                }
                if !inSearch {
                    count += 1
                }
                if !CurrentUser.sharedInstance().isLoggedIn {
                    count += 1
                }
                return count
            }
            if entries.count > 0 {
                return tableData[section-1].count
            }
            return 0
        }
        
        override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
            return UILocalizedIndexedCollation.currentCollation().sectionTitles.count + 1
        }
        
        var badImages = [String]()
        
        override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
            if indexPath.section == 0 {
                switch (indexPath.row, CurrentUser.sharedInstance().isLoggedIn, inSearch, entries.count == 0) {
                case (0, true, false, _), (0, false, false, _):
                    return tableView.dequeueReusableCellWithIdentifier("Directory Initial Message Cell", forIndexPath: indexPath) as UITableViewCell
                case (0, false, true, _), (1, false, false, _):
                    return tableView.dequeueReusableCellWithIdentifier("Directory Sign In Message Cell", forIndexPath: indexPath) as UITableViewCell
                case (0, true, true, true), (1, false, true, true):
                    return tableView.dequeueReusableCellWithIdentifier("Directory Empty Message Cell", forIndexPath: indexPath) as UITableViewCell
                default:
                    return UITableViewCell()
                }
            } else {
                let entry = self.tableData[indexPath.section-1][indexPath.row]
                let cell : UITableViewCell
                if let logo = entry.imageUrl where logo != "" {
                    if badImages.contains(logo) {
                        cell  = tableView.dequeueReusableCellWithIdentifier("Directory Name Cell", forIndexPath: indexPath) as UITableViewCell
                    } else {
                        cell  = tableView.dequeueReusableCellWithIdentifier("Directory Name Image Cell", forIndexPath: indexPath) as UITableViewCell
                        let imageView = cell.viewWithTag(3) as! UIImageView
                        imageView.loadImagefromURL(logo, successHandler: {
                            imageView.hidden = false
                            imageView.convertToCircleImage() }, failureHandler:  {
                            dispatch_async(dispatch_get_main_queue()) {
                                () -> Void in
                                imageView.hidden = true
                                self.badImages.append(logo)
                                self.tableView.reloadRowsAtIndexPaths([indexPath], withRowAnimation: .None)
                            }
                            
                            
                            }
                        )
                    }
                } else {
                    cell  = tableView.dequeueReusableCellWithIdentifier("Directory Name Cell", forIndexPath: indexPath) as UITableViewCell
                }
                
                let nameLabel = cell.viewWithTag(1) as! UILabel
                let typeLabel = cell.viewWithTag(2) as! UILabel
                nameLabel.text = entry.nameOrderedByFirstName(!self.sortByLastName)
                typeLabel.text = formatType(entry)
                return cell
            }
        }
        
        func formatType(entry: DirectoryEntry) -> String? {
            if let type = entry.type {
                if !legacySearch {
                    return type
                } else {
                    switch type {
                    case "student": return NSLocalizedString("Students", comment: "student search scope in directory")
                    case "faculty": return NSLocalizedString("Faculty/Staff", comment:"facilty/staff search scope in directory")
                    default: return type
                        
                    }
                }
                
            }
            return nil
        }
        
        override func tableView(tableView: UITableView, willSelectRowAtIndexPath indexPath: NSIndexPath) -> NSIndexPath? {
            return indexPath.section == 0 ? nil : indexPath
        }
        
        override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
            let entry = self.tableData[indexPath.section-1][indexPath.row]
            self.performSegueWithIdentifier("Show Directory Profile", sender: entry)
        }
        
        func updateFilter(filter: [String]) {
            if !self.splitViewController!.collapsed {
                self.performSegueWithIdentifier("Empty Entry", sender: nil)
            }
            self.hiddenGroups = filter
            if groups.count == filter.count {
                clear()
            } else {
                doSearch()
            }
        }
        
        func signInHappened() {
            dispatch_async(dispatch_get_main_queue(),{
                self.tableView.reloadData()
            })
        }
        
        func buildGroups () {
            
            
            let moduleKey = "\(module!.internalKey!)-hiddenGroups"
            if let hiddenGroups = AppGroupUtilities.userDefaults()?.arrayForKey(moduleKey) as? [String] {
                self.hiddenGroups = hiddenGroups
            } else {
                self.hiddenGroups = [String]()
            }
            
            do {
                let fetchRequest = NSFetchRequest(entityName: "DirectoryDefinition")
                fetchRequest.sortDescriptors = [ NSSortDescriptor(key: "displayName", ascending: true)]
                
                if let groups = try self.module?.managedObjectContext?.executeFetchRequest(fetchRequest) as? [DirectoryDefinition] where groups.count > 0 && module?.propertyForKey("directories") != nil {
                    let supportedDirectories = module?.propertyForKey("directories").componentsSeparatedByString(",")
                    let moduleGroups = groups.filter() {
                        supportedDirectories!.contains($0.key!)
                    }
                    self.groups = moduleGroups
                } else {
                    legacySearch = true
                    if self.module?.propertyForKey("student") == "true" && self.module?.propertyForKey("faculty") == "true" {
                        self.groups = [ legacyStudentDefinition(), legacyFacultyDefinition() ]
                    } else if self.module?.propertyForKey("student") == "true" {
                        self.groups = [ legacyStudentDefinition() ]
                    } else if self.module?.propertyForKey("faculty") == "true" {
                        self.groups = [ legacyFacultyDefinition() ]
                    } else {
                        self.groups = [ legacyStudentDefinition(), legacyFacultyDefinition() ]
                    }
                    self.groups = self.groups.sort( { $0.displayName < $1.displayName})
                }
            } catch {
            }
        }
        
        func legacyStudentDefinition() -> DirectoryDefinitionProtocol {
            let dir = DirectoryUnmanagedDefinition()
            dir.internalName = "student"
            dir.displayName = NSLocalizedString("Students", comment: "student search scope in directory")
            dir.authenticatedOnly = true
            return dir
        }
        
        func legacyFacultyDefinition() -> DirectoryDefinitionProtocol {
            let dir = DirectoryUnmanagedDefinition()
            dir.internalName = "faculty"
            dir.displayName = NSLocalizedString("Faculty/Staff", comment:"facilty/staff search scope in directory")
            dir.authenticatedOnly = true
            return dir
        }
    }